src/loaders/genome_collection/parse_computed_genome_attribs.py

```diff
@@ -133,16 +192,366 @@ def _read_tool_result(result_dir, batch_dir, kbase_collection, load_ver, tool_fi
+
+
+def _find_trait_by_index(
+        trait_index: int,
+        trait_meta: dict[str, dict[str, int | str]]
+) -> dict[str, int | str]:
+    # Find a trait in the global traits dictionary by its ID
+    for trait in trait_meta.values():
+        if trait[_SYS_TRAIT_INDEX] == trait_index:
+            return trait
+
+    raise ValueError(f'Unable to find trait with ID {trait_index}')
+
+
+def _ensure_list_ordered(a_list: list[str]) -> bool:
+    # Given a list of int strings, check if the list is ordered in ascending order
+    return a_list == sorted(a_list, key=int)
+
+
+def _parse_heatmap_rows(
+        traits_meta: dict[str, dict[str, int | str]],
+        traits_val: dict[str, list[dict[str, int | float | bool]]],
+) -> (list[HeatMapRow], float | int, float | int):
+    min_value, max_value, cell_count = float('inf'), float('-inf'), 0
+    heatmap_rows, trait_idxs = [], set([trait[_SYS_TRAIT_INDEX] for trait in traits_meta.values()])
+    for data_id, traits_val_list in traits_val.items():
+        heatmap_row = HeatMapRow(kbase_id=data_id, cells=[])
+        visited_traits = set()
+        for trait_val_info in traits_val_list:
+            trait_idx = trait_val_info[_SYS_TRAIT_INDEX]
+            trait_val = trait_val_info[_SYS_TRAIT_VALUE]
+            trait = _find_trait_by_index(trait_idx, traits_meta)
+            trait_type = trait.get(_SYS_TRAIT_TYPE)
+
+            min_value, max_value = _append_cell(heatmap_row,
+                                                trait_idx,
+                                                cell_count,
+                                                trait_type,
+                                                min_value,
+                                                max_value,
+                                                trait_val=trait_val)
+            visited_traits.add(trait_idx)
+            cell_count += 1
+
+        # fill in missing trait values with 0s for all cells
+        missing_trait_idxs = trait_idxs - visited_traits
+        for missing_trait_idx in missing_trait_idxs:
+            trait = _find_trait_by_index(missing_trait_idx, traits_meta)
+            trait_type = trait.get(_SYS_TRAIT_TYPE)
+            min_value, max_value = _append_cell(heatmap_row,
+                                                missing_trait_idx,
+                                                cell_count,
+                                                trait_type,
+                                                min_value,
+                                                max_value,
+                                                trait_val=_SYS_DEFAULT_TRAIT_VALUE)
+            cell_count += 1
+
+        # sort the cells by column ID to ensure the heatmap is in the correct order
+        heatmap_row.cells = sorted(heatmap_row.cells, key=lambda cell: int(cell.colid))
+        cell_col_ids = [cell.colid for cell in heatmap_row.cells]
+        if not _ensure_list_ordered(cell_col_ids):
```

Comment:
I don't see how this could ever happen, since the sort by column ID happens right above

Reply:
:+1: